keps/0010-dynamic-plugin-configuration.md

```diff
@@ -0,0 +1,221 @@
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+Currently the configuration for the `KongPlugin` and `KongClusterPlugin` resources
+is generally static, meaning that it is not possible to derive configuration values from
+common configuration resources defined on same kubernetes cluster. This leads to a lot
+duplication of hard coded values in numerous places making the deployment process
+difficult to manage. While it is possible to load configuration from a secret using
+the `configFrom` parameter, the entire configuration must be contained with in the
+secret which is effectively the same problem.
+
+It would be advantageous, and more in line with common Kubernetes behavior to be able
+to derive individual values of plugin configuration from data already stored on the same
+Kubernetes cluster. This proposal aims to define a change to the `KongPlugin` and `KongClusterPlugin`
+Custom Resource Definitions to facilitate this behavior.
+
+## Motivation
+
+Allow `KongPlugin` and `KongClusterPlugin` resources to read configuration values from Kubernetes,
+in addition to literal values.
+
+Continue to allow `KongPlugin` and `KongClusterPlugin` resource to accept literal input values.
+
+Reduce the amount of duplication of configuration value in the simple case. That is,
+when using standard kubernetes objects and with out assumed templating systems.
+
+### Goals
+
+* Provide ability to read individual configuration values from `ConfigMap` resources
+* Provide ability to read individual configuration values from `Secret` resources
+* Provide ability to read individual configuration values from arbitrary or generic resource types (other custom resource types)
+
+### Non-Goals
+
+* Introduction of a new Custom Resource or different Plugin Resource
+
+## Proposal
+
+In an effort to stay in line with current Kubernetes resource patterns it would be desirable
+to mimic the way pods define environment variables using an array of objects where
+the objects contain a `name` and one of `value` or `valueFrom`. `valueFrom`
+would further supports `configMapKeyRef`, `secretKeyRef` or `genericKeyRef`
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a operator I want certain plugin configuration values to be dynamic in addition
+to statically definable
+
+#### Story 2
+
+As a operator I want to be able to derive a literal value using value interpolation referencing
+other values
+
+#### Story 3
+
+As an operator I want values stored in `ConfigMap` or `Secret` resources
+to be the source of truth for various resources deployed to my Kubernetes cluster
+
+#### Story 4
+
+As an operator I expect changes in `ConfigMap` or `Secret` data values to be reflected
+in the resource and deployments that depend on them with out having to restart critical parts of the system
+
+#### Story 5
+
+As an operator I expect optional dynamic values to support default values to be utilized
+when the key or resource is missing from the cluster
```

Comment:
I'm sorry to say it's not clear to me what this story intends. Is this suggesting that when someone makes a reference to a `ConfigMap` or a `Secret` but that reference can't be fufilled we have backup configuration options which try to fill in the missing pieces? I could see this as being potentially quite difficult and maybe hazardous if we try to do it in an automagical kind of way :thinking:

Reply:
Yes, that.
But, not magic, explicit.

```yaml
configFrom:
  - name: FOO
    valueFrom:
      configMapKeyRef:
        name: my-secret
        key: foo
        optional: true
        default: 100   # <-----  a sane default
        type: number
```

Reply:
Ah I see that example makes perfect sense, thank you. Perhaps this wording could help:

```suggestion
As an operator I expect to have the option to provide default values for plugin configuration which
will be used in the case that a referent object is not available or readable.
```

Let me know your thoughts? :thinking:

Reply:
I like it. Done